src/plots/cartesian/axes.js

```diff
@@ -2208,26 +2221,39 @@ axes.drawGrid = function(gd, ax, opts) {
     opts = opts || {};
 
     var cls = ax._id + 'grid';
+    var vals = opts.vals;
+    var counterAx = opts.counterAxis;
+    if(ax.showgrid === false) {
+        vals = [];
+    }
+    else if(counterAx && axes.shouldShowZeroLine(gd, ax, counterAx)) {
```

Comment:
Does this fix any bugs, or just a nicer way to filter grid lines on top of zero lines?

Reply:
hmm, I thought it fixed some obscure bug but I'm not seeing what it is now. But mostly I just wanted to get away from the pattern of "create extras and then delete them," and 🌴 the logic of `axes.shouldShowZeroLine`

Reply:
Ok. No need to dig deeper, I was just curious. I like this pattern much better too!